compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -539,6 +536,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
 
             let (formal_input_ty, expected_input_ty) = formal_and_expected_inputs[expected_idx];
+            let formal_input_ty = self.resolve_vars_if_possible(formal_input_ty);
```

Comment:
Can you explain why you moved these resolves from the creation of `formal_and_expected_inputs`? It seems that a single call to resolve is now multiplied into very many, and it's not clear why.

Reply:
This is sort of the crux of the work, though I was far from thrilled at littering `self.resolve_vars_if_possible` all over. Once that is called in the method in the method above you lose the information about the generics they come from. In that case I can't later on do checks for "did another parameter that needed to match type T provide a different type to the current type".

So in the example provided `foo(a: T, b: T)` called with `foo(1, 2.0)` we know from `formal_and_expected_tys` that `a` has both formal_ty and expected_ty `TyVar(Int)` where `b` has formal_ty `TyVar(Int)` and expected_var `TyVar(Float)` but you don't know from the variables passed in that they both had their specified type of the same generic `T`. Moving the `self.resolve_vars_if_possible` at the level above made this information available.

Other options I can see:
* Could try and get this back out of the HIR (this seems reasonable but didn't think of it at the time)
* Could add another variable at the top of this function `formal_and_expected_tys_resolved` or something like that. I half had this in mind when writing this but wanted to see if this worked at the time so ploughed on.

NB: What is seemingly the case is I can absolutely change L512/509 back to original, this I use nowhere so reverted that back.

Reply:
I think the latest changes are better, they've removed all this and take it out of the HIR now.